### PR TITLE
[OCPCLOUD-930]: Aggregate operator errors before failing to sync

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Previously if one resource failed to reconcile mao would break the loop. Now it should try resyncing all resources and aggregate errors before failing and break the loop.

### CODE REVIEW

File: pkg/operator/sync.go [hunk 1]
```diff
@@ -56,53 +57,80 @@ func (optr *Operator) syncAll(config *OperatorConfig) error {
 		return nil
 	}
 
+	errors := []error{}
 	// Sync webhook configuration
 	if err := optr.syncWebhookConfiguration(); err != nil {
+		errors = append(errors, fmt.Errorf("Error syncing machine API webhook configurations: %w", err))
```
Comment (michaelgugino) [inline pkg/operator/sync.go:63]: The problem with this is we're increasing the likelihood that webhooks aren't running on first install (maybe other scenarios?) prior to the machine-api controllers being deployed.  This is going to cause bad things to happen down the road.

File: pkg/operator/sync.go [hunk 2]
```diff
@@ -56,53 +57,80 @@ func (optr *Operator) syncAll(config *OperatorConfig) error {
...
+
+	if err := optr.syncClusterAPIController(config); err != nil {
+		errors = append(errors, fmt.Errorf("Error syncing machine-api-controller: %w", err))
+	}
+
+	// Sync Termination Handler DaemonSet if supported
+	if config.Controllers.TerminationHandler != clusterAPIControllerNoOp {
+		if err := optr.syncTerminationHandler(config); err != nil {
+			errors = append(errors, fmt.Errorf("Error syncing termination handler: %w", err))
+			return err
```
Comment (enxebre) [inline pkg/operator/sync.go:74]: did you meant to return early here?

File: pkg/operator/sync.go [hunk 3]
```diff
@@ -56,53 +57,80 @@ func (optr *Operator) syncAll(config *OperatorConfig) error {
...
+		}
+	}
+
+	// In addition, if the Provider is BareMetal, then bring up
```
Comment (enxebre) [inline pkg/operator/sync.go:78]: where is this check?

File: pkg/operator/sync.go [hunk 4]
```diff
@@ -56,53 +57,80 @@ func (optr *Operator) syncAll(config *OperatorConfig) error {
...
 			glog.Errorf("Error syncing ClusterOperatorStatus: %v", err)
 		}
-		glog.Errorf("Error syncing machine API webhook configurations: %v", err)
+		glog.Errorf("Error syncing machine controller components: %v", err)
 		return err
 	}
-	glog.V(3).Info("Synced up all machine API webhook configurations")
 
-	if err := optr.syncClusterAPIController(config); err != nil {
+	glog.V(3).Info("Synced up all machine controller components")
```
Comment (enxebre) [inline pkg/operator/sync.go:96]: this would only be true after the rollout right?

File: pkg/operator/sync.go (hunk 1 above)
Comment (enxebre) [reply pkg/operator/sync.go:63]: yeh webhooks should fail early

**Comment by enxebre** [comment]
thanks @Danil-Grigorev Going through the changes now is not clear we are actually taken any advantage out of it. Let's punt on this for now and reevaluate later in the new release.

**Comment by Danil-Grigorev** [comment]
/hold cancel
It is ready to merge in 4.7

File: pkg/operator/sync.go (hunk 2 above)
Comment (elmiko) [reply pkg/operator/sync.go:79]: this doesn't make sense to me also, any clarification here Danil?

**Review by elmiko — COMMENTED** [review]
this mostly makes sense, but i am curious about the question Alberto raised

**Comment by Danil-Grigorev** [comment]
Good catch @elmiko, thanks... Fixed now - https://github.com/openshift/machine-api-operator/pull/691#discussion_r537778976

**Review by elmiko — APPROVED** [review]
/lgtm

**Comment by JoelSpeed** [comment]
/hold

@Danil-Grigorev If you want to get this into 4.7, it will need a bug associated. I'm not sure whether we need to now or whether we should hold this until 4.8.

**Comment by Danil-Grigorev** [comment]
No objection to hold until 4.8.
/release 4.8

**Pull request merged**